### PR TITLE
Implement ARM AAPCS ABI and other ARM fixes

### PR DESCRIPTION
**Pull request by smolt**
This is the bulk of the AAPCS (Procedure Call Standard for the ARM Architecture) used for C interop.  This will be used for both arm and thumb target triples.  The extern(D) ABI tries to follows the AAPCS with a few exceptions.

### CODE REVIEW

**Comment by smolt** [comment]
Expanding with fix for relocation model.  Now -O0 works.

Clang makes it an error if dynamic-no-pic is used with something other than darwin.  Should LDC do that too?  What do you think.

File: gen/abi.cpp [hunk 1]
```diff
@@ -235,6 +343,9 @@ TargetABI *TargetABI::getTarget() {
   case llvm::Triple::aarch64:
   case llvm::Triple::aarch64_be:
     return getAArch64TargetABI();
+  case llvm::Triple::arm:
+  case llvm::Triple::thumb:
```
Comment (redstar) [inline gen/abi.cpp]: What about `armeb` and `thumbeb`?

**Comment by redstar** [comment]
Never saw this many unit tests pass on my IFC6410. Really great!

File: gen/abi.cpp (hunk 1 above)
Comment (smolt) [inline gen/abi.cpp]: Yeah, I almost always forget them.

**Comment by smolt** [comment]
@redstar, do you have list of what failed?  I am going to open an issue later to work off remaining arm-linux failures (not many) unless you beat me to it.  We can cross check our failure lists.

**Pull request merged**